auth/src/main/java/com/firebase/ui/auth/util/data/AnonymousUpgradeUtils.java

```diff
@@ -0,0 +1,80 @@
+package com.firebase.ui.auth.util.data;
+
+import android.content.Context;
+import android.content.Intent;
+import android.support.annotation.NonNull;
+import android.support.annotation.RestrictTo;
+
+import com.firebase.ui.auth.ErrorCodes;
+import com.firebase.ui.auth.IdpResponse;
+import com.firebase.ui.auth.data.model.FlowParameters;
+import com.firebase.ui.auth.util.ExtraConstants;
+import com.google.android.gms.tasks.Task;
+import com.google.firebase.FirebaseApp;
+import com.google.firebase.auth.AuthCredential;
+import com.google.firebase.auth.AuthResult;
+import com.google.firebase.auth.EmailAuthProvider;
+import com.google.firebase.auth.FirebaseAuth;
+
+/**
+ * Utilities to help with Anonymous user upgrade.
+ */
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+public class AnonymousUpgradeUtils {
+
+    private static String firebaseAppName = "ScratchApp";
+
+    public static synchronized FirebaseApp getFirebaseApp(Context context) {
+        try {
+            FirebaseApp scratchApp = FirebaseApp.getInstance(firebaseAppName);
+            return scratchApp;
+        } catch (IllegalStateException e) {
+            FirebaseApp app = FirebaseApp.getInstance();
+            return FirebaseApp.initializeApp(context, app.getOptions(), firebaseAppName);
+        }
+    }
+
+    public static Task<AuthResult> createOrLinkUserWithEmailAndPassword(@NonNull FirebaseAuth auth,
+                                                                        @NonNull FlowParameters flowParameters,
+                                                                        @NonNull String email,
+                                                                        @NonNull String password) {
+        if (canUpgradeAnonymous(auth, flowParameters)) {
+            AuthCredential credential = EmailAuthProvider.getCredential(email, password);
+            return auth.getCurrentUser().linkWithCredential(credential);
+        } else {
+            return auth.createUserWithEmailAndPassword(email, password);
+        }
+    }
+
+    public static Task<AuthResult> signInAndLinkWithPhoneNumber(@NonNull FirebaseAuth auth,
```

Comment:
nit: I think you mean `signInOrLink`, right?

Reply:
Well the desired outcome is that the link is successful, which will result in the account being linked and a sign in. If the link fails then we don't proceed with the sign in, the dev does.